src/FuseGPUThreadLoops.cpp

```diff
@@ -347,51 +381,93 @@ class ExtractSharedAllocations : public IRMutator {
     }
 
     Stmt visit(const Allocate *op) override {
-        user_assert(!op->new_expr.defined()) << "Allocate node inside GPU kernel has custom new expression.\n"
-                                             << "(Memoization is not supported inside GPU kernels at present.)\n";
-
-        if (in_threads ||
-            op->memory_type == MemoryType::Stack ||
-            op->memory_type == MemoryType::Register) {
-            // TODO: Support shared allocations inside loops over threads by giving each loop iteration a unique slice.
+        user_assert(!op->new_expr.defined())
+            << "Allocate node inside GPU kernel has custom new expression.\n"
+            << "(Memoization is not supported inside GPU kernels at present.)\n";
+
+        bool fixed_size_thread_allocation = (op->constant_allocation_size() != 0) && in_threads;
+
+        if ((fixed_size_thread_allocation &&
+             op->memory_type != MemoryType::Heap &&
+             op->memory_type != MemoryType::GPUShared) ||
+            op->memory_type == MemoryType::Register ||
```

Comment:
Alignment is weird here (clang-format issue?)

Reply:
Alignment seems legit to me - the structure of the expression is (A && B && C) || D || E, so A, B, C, are one space further in to line up with the open paren